File: libraries/flicker/src/com/android/server/wm/flicker/TransitionRunner.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.server.wm.flicker;

import static com.android.server.wm.flicker.monitor.TransitionMonitor.OUTPUT_DIR;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.test.InstrumentationRegistry;

import com.android.server.wm.flicker.monitor.TransitionMonitor;
import com.android.server.wm.flicker.monitor.LayersTraceMonitor;
import com.android.server.wm.flicker.monitor.ScreenRecorder;
import com.android.server.wm.flicker.monitor.WindowAnimationFrameStatsMonitor;
import com.android.server.wm.flicker.monitor.WindowManagerTraceMonitor;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;

/**
 * Builds and runs UI transitions capturing test artifacts.
 *
 * <p>User can compose a transition from simpler steps, specifying setup and teardown steps. During
 * a transition, Layers trace, WindowManager trace, screen recordings and window animation frame
 * stats can be captured.
 *
 * <pre>
 * Transition builder options:
 *  {@link TransitionBuilder#run(Runnable)} run transition under test. Monitors will be started
 *  before the transition and stopped after the transition is completed.
 *  {@link TransitionBuilder#repeat(int)} repeat transitions under test multiple times recording
 *  result for each run.
 *  {@link TransitionBuilder#withTag(String)} specify a string identifier used to prefix logs and
 *  artifacts generated.
 *  {@link TransitionBuilder#runBeforeAll(Runnable)} run setup transitions once before all other
 *  transition are run to set up an initial state on device.
 *  {@link TransitionBuilder#runBefore(Runnable)} run setup transitions before each test transition
 *  run.
 *  {@link TransitionBuilder#runAfter(Runnable)} run teardown transitions after each test
 *  transition.
 *  {@link TransitionBuilder#runAfter(Runnable)} run teardown transitions once after all
 *  other transition  are run.
 *  {@link TransitionBuilder#includeJankyRuns()} disables {@link WindowAnimationFrameStatsMonitor}
 *  to monitor janky frames. If janky frames are detected, then the test run is skipped. This
 *  monitor is enabled by default.
 *  {@link TransitionBuilder#skipLayersTrace()} disables {@link LayersTraceMonitor} used to
 *  capture Layers trace during a transition. This monitor is enabled by default.
 *  {@link TransitionBuilder#skipWindowManagerTrace()} disables {@link WindowManagerTraceMonitor}
 *  used to capture WindowManager trace during a transition. This monitor is enabled by
 *  default.
 *  {@link TransitionBuilder#recordAllRuns()} records the screen contents and saves it to a file.
 *  All the runs including setup and teardown transitions are included in the recording. This
 *  monitor is used for debugging purposes.
 *  {@link TransitionBuilder#recordEachRun()} records the screen contents during test transitions
 *  and saves it to a file for each run. This monitor is used for debugging purposes.
 *
 * Example transition to capture WindowManager and Layers trace when opening a test app:
 * {@code
 * TransitionRunner.newBuilder()
 *      .withTag("OpenTestAppFast")
 *      .runBeforeAll(UiAutomationLib::wakeUp)
 *      .runBeforeAll(UiAutomationLib::UnlockDevice)
 *      .runBeforeAll(UiAutomationLib::openTestApp)
 *      .runBefore(UiAutomationLib::closeTestApp)
 *      .run(UiAutomationLib::openTestApp)
 *      .runAfterAll(UiAutomationLib::closeTestApp)
 *      .repeat(5)
 *      .build()
 *      .run();
 * }
 * </pre>
 */
public class TransitionRunner {
    private static final String TAG = "FLICKER";
    private final ScreenRecorder mScreenRecorder;
    private final WindowManagerTraceMonitor mWmTraceMonitor;
    private final LayersTraceMonitor mLayersTraceMonitor;
    private final WindowAnimationFrameStatsMonitor mFrameStatsMonitor;

    private final List<TransitionMonitor> mAllRunsMonitors;
    private final List<TransitionMonitor> mPerRunMonitors;
    private final List<Runnable> mBeforeAlls;
    private final List<Runnable> mBefores;
    private final List<Runnable> mTransitions;
    private final List<Runnable> mAfters;
    private final List<Runnable> mAfterAlls;

    private final int mIterations;
    private final String mTestTag;

    @Nullable private List<TransitionResult> mResults = null;

    private TransitionRunner(TransitionBuilder builder) {
        mScreenRecorder = builder.mScreenRecorder;
        mWmTraceMonitor = builder.mWmTraceMonitor;
        mLayersTraceMonitor = builder.mLayersTraceMonitor;
        mFrameStatsMonitor = builder.mFrameStatsMonitor;

        mAllRunsMonitors = builder.mAllRunsMonitors;
        mPerRunMonitors = builder.mPerRunMonitors;
        mBeforeAlls = builder.mBeforeAlls;
        mBefores = builder.mBefores;
        mTransitions = builder.mTransitions;
        mAfters = builder.mAfters;
        mAfterAlls = builder.mAfterAlls;

        mIterations = builder.mIterations;
        mTestTag = builder.mTestTag;
    }

    public static TransitionBuilder newBuilder(String outputDir) {
        return new TransitionBuilder(Paths.get(outputDir));
    }

    public static TransitionBuilder newBuilder() {
        return new TransitionBuilder();
    }

    /**
     * Runs the composed transition and calls monitors at the appropriate stages. If jank monitor is
     * enabled, transitions with jank are skipped.
     *
     * @return itself
     */
    public TransitionRunner run() {
        mResults = new ArrayList<>();
        mAllRunsMonitors.forEach(TransitionMonitor::start);
        mBeforeAlls.forEach(Runnable::run);
        for (int iteration = 0; iteration < mIterations; iteration++) {
            mBefores.forEach(Runnable::run);
            mPerRunMonitors.forEach(TransitionMonitor::start);
            mTransitions.forEach(Runnable::run);
            mPerRunMonitors.forEach(TransitionMonitor::stop);
            mAfters.forEach(Runnable::run);
            if (runJankFree() && mFrameStatsMonitor.jankyFramesDetected()) {
                String msg =
                        String.format(
                                Locale.getDefault(),
                                "Skipping iteration %d/%d for test %s due to jank. %s",
                                iteration,
                                mIterations - 1,
                                mTestTag,
                                mFrameStatsMonitor.toString());
                Log.e(TAG, msg);
                continue;
            }
            mResults.add(saveResult(iteration));
        }
        mAfterAlls.forEach(Runnable::run);
        mAllRunsMonitors.forEach(
                monitor -> {
                    monitor.stop();
                    monitor.save(mTestTag);
                });
        return this;
    }

    /**
     * Returns a list of transition results.
     *
     * @return list of transition results.
     */
    public List<TransitionResult> getResults() {
        if (mResults == null) {
            throw new IllegalStateException("Results do not exist!");
        }
        return mResults;
    }

    /**
     * Deletes all transition results that are not marked for saving.
     *
     * @return list of transition results.
     */
    public void deleteResults() {
        if (mResults == null) {
            return;
        }
        mResults.stream().filter(TransitionResult::canDelete).forEach(TransitionResult::delete);
        mResults = null;
    }

    /**
     * Saves monitor results to file.
     *
     * @return object containing paths to test artifacts
     */
    private TransitionResult saveResult(int iteration) {
        Path windowTrace = null;
        String windowTraceChecksum = "";
        Path layerTrace = null;
        String layerTraceChecksum = "";
        Path screenCaptureVideo = null;
        String screenCaptureVideoChecksum = "";

        if (mPerRunMonitors.contains(mWmTraceMonitor)) {
            windowTrace = mWmTraceMonitor.save(mTestTag, iteration);
            windowTraceChecksum = mWmTraceMonitor.getChecksum();
        }
        if (mPerRunMonitors.contains(mLayersTraceMonitor)) {
            layerTrace = mLayersTraceMonitor.save(mTestTag, iteration);
            layerTraceChecksum = mLayersTraceMonitor.getChecksum();
        }
        if (mPerRunMonitors.contains(mScreenRecorder)) {
            screenCaptureVideo = mScreenRecorder.save(mTestTag, iteration);
            screenCaptureVideoChecksum = mScreenRecorder.getChecksum();
        }
        return new TransitionResult(
                layerTrace,
                layerTraceChecksum,
                windowTrace,
                windowTraceChecksum,
                screenCaptureVideo,
                screenCaptureVideoChecksum);
    }

    private boolean runJankFree() {
        return mPerRunMonitors.contains(mFrameStatsMonitor);
    }

    public String getTestTag() {
        return mTestTag;
    }

    /** Stores paths to all test artifacts. */
    @VisibleForTesting
    public static class TransitionResult {
        @Nullable private final Path layersTrace;
        private final String layersTraceChecksum;
        @Nullable private final Path windowManagerTrace;
        private final String windowManagerTraceChecksum;
        @Nullable private final Path screenCaptureVideo;
        private final String screenCaptureVideoChecksum;
        private boolean flaggedForSaving = true;

        public TransitionResult(
                @Nullable Path layersTrace,
                String layersTraceChecksum,
                @Nullable Path windowManagerTrace,
                String windowManagerTraceChecksum,
                @Nullable Path screenCaptureVideo,
                String screenCaptureVideoChecksum) {
            this.layersTrace = layersTrace;
            this.layersTraceChecksum = layersTraceChecksum;
            this.windowManagerTrace = windowManagerTrace;
            this.windowManagerTraceChecksum = windowManagerTraceChecksum;
            this.screenCaptureVideo = screenCaptureVideo;
            this.screenCaptureVideoChecksum = screenCaptureVideoChecksum;
        }

        public void flagForSaving() {
            flaggedForSaving = true;
        }

        public boolean canDelete() {
            return !flaggedForSaving;
        }

        public boolean layersTraceExists() {
            return layersTrace != null && layersTrace.toFile().exists();
        }

        public byte[] getLayersTrace() {
            try {
                return Files.readAllBytes(this.layersTrace);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }

        public Path getLayersTracePath() {
            return layersTrace;
        }

        public String getLayersTraceChecksum() {
            return layersTraceChecksum;
        }

        public boolean windowManagerTraceExists() {
            return windowManagerTrace != null && windowManagerTrace.toFile().exists();
        }

        public byte[] getWindowManagerTrace() {
            try {
                return Files.readAllBytes(this.windowManagerTrace);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }

        public Path getWindowManagerTracePath() {
            return windowManagerTrace;
        }

        public String getWindowManagerTraceChecksum() {
            return windowManagerTraceChecksum;
        }

        public boolean screenCaptureVideoExists() {
            return screenCaptureVideo != null && screenCaptureVideo.toFile().exists();
        }

        public Path screenCaptureVideoPath() {
            return screenCaptureVideo;
        }

        public String getScreenCaptureVideoChecksum() {
            return screenCaptureVideoChecksum;
        }

        public void delete() {
            if (layersTraceExists()) layersTrace.toFile().delete();
            if (windowManagerTraceExists()) windowManagerTrace.toFile().delete();
            if (screenCaptureVideoExists()) screenCaptureVideo.toFile().delete();
        }
    }

    /** Builds a {@link TransitionRunner} instance. */
    public static class TransitionBuilder {
        private ScreenRecorder mScreenRecorder;
        private WindowManagerTraceMonitor mWmTraceMonitor;
        private LayersTraceMonitor mLayersTraceMonitor;
        private WindowAnimationFrameStatsMonitor mFrameStatsMonitor;

        private List<TransitionMonitor> mAllRunsMonitors = new LinkedList<>();
        private List<TransitionMonitor> mPerRunMonitors = new LinkedList<>();
        private List<Runnable> mBeforeAlls = new LinkedList<>();
        private List<Runnable> mBefores = new LinkedList<>();
        private List<Runnable> mTransitions = new LinkedList<>();
        private List<Runnable> mAfters = new LinkedList<>();
        private List<Runnable> mAfterAlls = new LinkedList<>();

        private boolean mRunJankFree = true;
        private boolean mCaptureWindowManagerTrace = true;
        private boolean mCaptureLayersTrace = true;
        private boolean mRecordEachRun = false;
        private int mIterations = 1;
        private String mTestTag = "";

        private boolean mRecordAllRuns = false;

        private TransitionBuilder(@NonNull Path outputDir) {
            mScreenRecorder = new ScreenRecorder();
            mWmTraceMonitor = new WindowManagerTraceMonitor(outputDir);
            mLayersTraceMonitor = new LayersTraceMonitor(outputDir);
            mFrameStatsMonitor =
                    new WindowAnimationFrameStatsMonitor(
                            InstrumentationRegistry.getInstrumentation());
        }

        public TransitionBuilder() {
            this(OUTPUT_DIR);
        }

        public TransitionRunner build() {
            if (mCaptureWindowManagerTrace) {
                mPerRunMonitors.add(mWmTraceMonitor);
            }

            if (mCaptureLayersTrace) {
                mPerRunMonitors.add(mLayersTraceMonitor);
            }

            if (mRunJankFree) {
                mPerRunMonitors.add(mFrameStatsMonitor);
            }

            if (mRecordAllRuns) {
                mAllRunsMonitors.add(mScreenRecorder);
            }

            if (mRecordEachRun) {
                mPerRunMonitors.add(mScreenRecorder);
            }

            return new TransitionRunner(this);
        }

        public TransitionBuilder runBeforeAll(Runnable runnable) {
            mBeforeAlls.add(runnable);
            return this;
        }

        public TransitionBuilder runBefore(Runnable runnable) {
            mBefores.add(runnable);
            return this;
        }

        public TransitionBuilder run(Runnable runnable) {
            mTransitions.add(runnable);
            return this;
        }

        public TransitionBuilder runAfter(Runnable runnable) {
            mAfters.add(runnable);
            return this;
        }

        public TransitionBuilder runAfterAll(Runnable runnable) {
            mAfterAlls.add(runnable);
            return this;
        }

        public TransitionBuilder repeat(int iterations) {
            mIterations = iterations;
            return this;
        }

        public TransitionBuilder skipWindowManagerTrace() {
            mCaptureWindowManagerTrace = false;
            return this;
        }

        public TransitionBuilder skipLayersTrace() {
            mCaptureLayersTrace = false;
            return this;
        }

        public TransitionBuilder includeJankyRuns() {
            mRunJankFree = false;
            return this;
        }

        public TransitionBuilder recordEachRun() {
            if (mRecordAllRuns) {
                throw new IllegalArgumentException("Invalid option with recordAllRuns");
            }
            mRecordEachRun = true;
            return this;
        }

        public TransitionBuilder recordAllRuns() {
            if (mRecordEachRun) {
                throw new IllegalArgumentException("Invalid option with recordEachRun");
            }
            mRecordAllRuns = true;
            return this;
        }

        public TransitionBuilder withTag(String testTag) {
            if (testTag.contains(" ")) {
                throw new IllegalArgumentException(
                        "The test tag can not contain spaces since it "
                                + "is a part of the file name");
            }
            mTestTag = testTag;
            return this;
        }
    }
}

File: libraries/flicker/src/com/android/server/wm/flicker/TransitionRunner.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.server.wm.flicker

import android.app.Instrumentation
import android.os.Bundle
import android.os.SystemClock
import android.util.Log
import androidx.test.platform.app.InstrumentationRegistry
import com.android.server.wm.flicker.FlickerRunResult.Companion.RunStatus
import com.android.server.wm.flicker.monitor.IFileGeneratingMonitor
import com.android.server.wm.flicker.monitor.ITransitionMonitor
import com.android.server.wm.flicker.monitor.NoTraceMonitor
import com.android.server.wm.traces.common.ConditionList
import com.android.server.wm.traces.common.WindowManagerConditionsFactory
import com.android.server.wm.traces.parser.getCurrentState
import java.io.IOException
import java.nio.file.Files
import java.util.concurrent.TimeUnit
import kotlin.system.measureTimeMillis
import org.junit.runner.Description

/**
 * Runner to execute the transitions of a flicker test
 *
 * The commands are executed in the following order:
 * 1) [Flicker.transitionSetup]
 * 2) Start monitors
 * 3) [Flicker.transitions]
 * 4) Stop monitors
 * 5) [Flicker.transitionTeardown]
 *
 * If the tests were already executed, reuse the previous results
 *
 */
open class TransitionRunner {
    private val tags = mutableSetOf<String>()
    private var _result: FlickerRunResult? = null
    private val result: FlickerRunResult get() {
        requireNotNull(_result) { "Result not initialized" }
        return _result!!
    }

    /**
     * Executes the setup, transitions and teardown defined in [flicker]
     *
     * @param flicker test specification
     * @throws IllegalArgumentException If the transitions are empty
     */
    open fun execute(flicker: Flicker, useCacheIfAvailable: Boolean = true): FlickerResult {
        check(flicker)
        return run(flicker)
    }

    /**
     * Validate the [flicker] test specification before executing the transitions
     *
     * @param flicker test specification
     * @throws IllegalArgumentException If the transitions are empty
     */
    protected fun check(flicker: Flicker) {
        require(flicker.transitions.isNotEmpty() || onlyHasNoTraceMonitors(flicker)) {
            "A flicker test must include transitions to run"
        }
    }

    private fun onlyHasNoTraceMonitors(flicker: Flicker) =
            flicker.traceMonitors.all { it is NoTraceMonitor }

    open fun cleanUp() {
        tags.clear()
        _result = null
    }

    /**
     * Runs the actual setup, transitions and teardown defined in [flicker]
     *
     * @param flicker test specification
     */
    internal open fun run(flicker: Flicker): FlickerResult {
        val runResult = FlickerRunResult(flicker.testName)
        _result = runResult
        safeExecution(flicker) {
            val description = Description.createSuiteDescription(flicker.testName)
            if (flicker.faasEnabled) {
                Log.d(FLICKER_TAG, "${flicker.testName} - Setting up FaaS")
                flicker.faas.testStarted(description)
            }

            progressUpdate("${flicker.testName} - Starting traces")
            measureTimeMillis {
                flicker.traceMonitors.forEach { it.start() }
            }.also {
                progressUpdate("\tTook ${it}ms to start traces")
            }


            progressUpdate("${flicker.testName} - Running transition setup")
            measureTimeMillis {
                runTransitionSetup(flicker)
            }.also {
                progressUpdate("\tTook ${it}ms to run transition setup")
            }

            progressUpdate("${flicker.testName} - Running transition")
            measureTimeMillis {
                runTransition(flicker)
            }.also {
                progressUpdate( "\tTook ${it}ms to run transition")
            }

            progressUpdate("${flicker.testName} - Running transition teardown")
            measureTimeMillis {
                runTransitionTeardown(flicker)
            }.also {
                progressUpdate("\tTook ${it}ms to run transition teardown")
            }

            progressUpdate("${flicker.testName} - Stopping traces")
            measureTimeMillis {
                flicker.traceMonitors.forEach { it.tryStop() }
            }.also {
                progressUpdate("\tTook ${it}ms to stop traces")
            }

            progressUpdate("${flicker.testName} - Processing transition traces")
            measureTimeMillis {
                processRunTraces(flicker, RunStatus.ASSERTION_SUCCESS)
            }.also {
                progressUpdate("\tTook ${it}ms to process traces")
            }

            if (flicker.faasEnabled) {
                progressUpdate("${flicker.testName} - " +
                        "Notifying FaaS of finished transition")
                flicker.faas.testFinished(description)
                if (flicker.faas.executionErrors.isNotEmpty()) {
                    runResult.setFaasExecutionError(flicker.faas.executionErrors[0])
                    for (executionError in flicker.faas.executionErrors) {
                        Log.e(FLICKER_TAG, "FaaS reported execution errors", executionError)
                    }
                }
            }
        }

        val result = FlickerResult(runResult)
        cleanUp()
        return result
    }

    private fun safeExecution(
        flicker: Flicker,
        alreadyFailed: Boolean = false,
        execution: () -> Unit
    ) {
        try {
            execution()
        } catch (e: ExecutionError) {
            Log.e(FLICKER_TAG, "A Flicker Execution Error occurred!", e)
            if (alreadyFailed) {
                // If we already failed don't try and handle failures since we are most likely
                // failing because of the previous execution error.
                Log.e(FLICKER_TAG, "FAILED HANDLING AN EXECUTION ERROR!")
                return
            }

            result.setTransitionExecutionError(e)
            result.setStatus(RunStatus.RUN_FAILED)

            when (e) {
                is TransitionSetupFailure,
                is TransitionExecutionFailure,
                is TransitionTeardownFailure -> {
                    // If we fail on the transition setup, the transition itself or the transition
                    // teardown, then we want to try to stop the tracing and store the traces in the
                    // trace archive.
                    safeExecution(flicker, true) {
                        Log.d(FLICKER_TAG, "${flicker.testName} - Stopping traces")
                        flicker.traceMonitors.forEach { it.tryStop() }

                        Log.d(FLICKER_TAG, "${flicker.testName} - Processing transition traces")
                        processRunTraces(flicker, RunStatus.RUN_FAILED)
                    }
                }
            }

            when (e) {
                is TransitionSetupFailure, is TransitionExecutionFailure -> {
                    // If we fail on the setup or transition we simply want to run the teardown.
                    safeExecution(flicker, true) {
                        runTransitionTeardown(flicker)
                    }
                }
                is TransitionTeardownFailure, is TraceProcessingFailure -> {
                    // Nothing to do & considered handled
                }
                else -> {
                    // Throw any unhandled error
                    throw e
                }
            }
        }
    }

    /**
     * Parses the traces collected by the monitors to generate FlickerRunResults containing the
     * parsed trace and information about the status of the run.
     * The run results are added to the resultBuilders list which is then used to run Flicker
     * assertions on.
     */
    @Throws(TraceProcessingFailure::class)
    private fun processRunTraces(
        flicker: Flicker,
        status: RunStatus
    ) {
        try {
            result.setStatus(status)
            setMonitorResults(flicker, result)
            result.lock()

            if (flicker.faasEnabled && !status.isFailure) {
                // Don't run FaaS on failed transitions
                val wmTrace = result.buildWmTrace()
                val layersTrace = result.buildLayersTrace()
                val transitionsTrace = result.buildTransitionsTrace()

                flicker.faasTracesCollector.wmTrace = wmTrace
                flicker.faasTracesCollector.layersTrace = layersTrace
                flicker.faasTracesCollector.transitionsTrace = transitionsTrace
            }
        } catch (e: Throwable) {
            // We have failed to add the results to the runs, so we can effectively consider these
            // results as "lost" as they won't be used from now forth. So we can safely rename
            // to file to indicate the failure and make it easier to find in the archives.
            flicker.traceMonitors.forEach {
                // All monitors that generate files we want to keep in the archives should implement
                // IFileGeneratingMonitor
                if (it is IFileGeneratingMonitor) {
                    Utils.addStatusToFileName(it.outputFile, RunStatus.PARSING_FAILURE)
                }
            }
            throw TraceProcessingFailure(e)
        }
    }

    @Throws(TransitionSetupFailure::class)
    private fun runTransitionSetup(flicker: Flicker) {
        try {
            flicker.transitionSetup.forEach { it.invoke(flicker) }
            flicker.wmHelper.StateSyncBuilder()
                .add(UI_STABLE_CONDITIONS)
                .waitFor()
        } catch (e: Throwable) {
            throw TransitionSetupFailure(e)
        }
    }

    @Throws(TransitionExecutionFailure::class)
    private fun runTransition(flicker: Flicker) {
        try {
            result.transitionStartTime = FlickerRunResult.TraceTime(
                elapsedRealtimeNanos = SystemClock.elapsedRealtimeNanos(),
                systemTime = SystemClock.uptimeNanos(),
                unixTimeNanos = TimeUnit.NANOSECONDS.convert(
                    System.currentTimeMillis(), TimeUnit.MILLISECONDS),
            )

            flicker.transitions.forEach { it.invoke(flicker) }
            flicker.wmHelper.StateSyncBuilder()
                    .add(UI_STABLE_CONDITIONS)
                    .waitFor()

            result.transitionEndTime = FlickerRunResult.TraceTime(
                elapsedRealtimeNanos = SystemClock.elapsedRealtimeNanos(),
                systemTime = SystemClock.uptimeNanos(),
                unixTimeNanos = TimeUnit.NANOSECONDS.convert(
                    System.currentTimeMillis(), TimeUnit.MILLISECONDS),
            )
        } catch (e: Throwable) {
            throw TransitionExecutionFailure(e)
        }
    }

    @Throws(TransitionTeardownFailure::class)
    private fun runTransitionTeardown(flicker: Flicker) {
        try {
            flicker.transitionTeardown.forEach { it.invoke(flicker) }
        } catch (e: Throwable) {
            throw TransitionTeardownFailure(e)
        }
    }

    private fun setMonitorResults(
        flicker: Flicker,
        result: FlickerRunResult,
    ): FlickerRunResult {

        flicker.traceMonitors.forEach {
            result.setResultsFromMonitor(it)
        }

        return result
    }

    private fun ITransitionMonitor.tryStop() {
        this.run {
            try {
                stop()
            } catch (e: Exception) {
                Log.e(FLICKER_TAG, "Unable to stop $this", e)
            }
        }
    }

    private fun getTaggedFilePath(flicker: Flicker, tag: String, file: String) =
        "${flicker.testName}_${tag}_$file"

    /**
     * Captures a snapshot of the device state and associates it with a new tag.
     *
     * This tag can be used to make assertions about the state of the device when the
     * snapshot is collected.
     *
     * [tag] is used as part of the trace file name, thus, only valid letters and digits
     * can be used
     *
     * @param flicker test specification
     * @throws IllegalArgumentException If [tag] contains invalid characters
     */
    open fun createTag(flicker: Flicker, tag: String) {
        require(!tag.contains(" ")) {
            "The test tag $tag can not contain spaces since it is a part of the file name"
        }
        tags.add(tag)

        val deviceStateBytes = getCurrentState(flicker.instrumentation.uiAutomation)
        try {
            val wmDumpFile = flicker.outputDir.resolve(
                getTaggedFilePath(flicker, tag, "wm_dump")
            )
            Files.write(wmDumpFile, deviceStateBytes.first)

            val layersDumpFile = flicker.outputDir.resolve(
                getTaggedFilePath(flicker, tag, "layers_dump")
            )
            Files.write(layersDumpFile, deviceStateBytes.second)

            result.addTaggedState(
                tag,
                wmDumpFile.toFile(),
                layersDumpFile.toFile(),
            )
        } catch (e: IOException) {
            throw RuntimeException("Unable to create trace file: ${e.message}", e)
        }
    }

    companion object {
        private fun progressUpdate(msg: String) {
            Log.d(FLICKER_TAG, msg)

            val instrumentation: Instrumentation = InstrumentationRegistry.getInstrumentation()
            val results = Bundle()
            results.putString(Instrumentation.REPORT_KEY_STREAMRESULT, "$msg\n")
            instrumentation.sendStatus(1, results)
        }

        /**
         * Conditions that determine when the UI is in a stable stable and no windows or layers are
         * animating or changing state.
         */
        private val UI_STABLE_CONDITIONS = ConditionList(
            listOf(
                WindowManagerConditionsFactory.isWMStateComplete(),
                WindowManagerConditionsFactory.hasLayersAnimating().negate()
            )
        )

        open class ExecutionError(private val inner: Throwable) : Throwable(inner) {
            init {
                super.setStackTrace(inner.stackTrace)
            }

            override val message: String?
                get() = inner.toString()
        }

        class TransitionSetupFailure(val e: Throwable) : ExecutionError(e)
        class TransitionExecutionFailure(val e: Throwable) : ExecutionError(e)
        class TraceProcessingFailure(val e: Throwable) : ExecutionError(e)
        class TransitionTeardownFailure(val e: Throwable) : ExecutionError(e)
    }
}
